lib/serializer.ts

```diff
@@ -554,6 +556,8 @@ function deserializeCompositeType(serializer: Serializer, mapper: CompositeMappe
         if (headerKey.startsWith(headerCollectionPrefix)) {
           dictionary[headerKey.substring(headerCollectionPrefix.length)] = serializer.deserialize((propertyMapper as DictionaryMapper).type.value, responseBody[headerKey], propertyObjectName);
         }
+
+        handledPropertyNames.push(headerKey);
```

Comment:
Adding handled properties in every path causes problems with nested properties with conflicting names.